Week5

jihyun-jeon opened the pull request:
5주차 코드리뷰 요청드립니다 :)

코테합(코딩테스트 합격자되기 책) 풀이 보며 개선사항도 파악해보았습니다.

jihyun-jeon committed:
feat: lv1.두개 뽑아서 더하기

jihyun-jeon committed:
feat: lv1.모의고사-완전탐색

jihyun-jeon committed:
feat: lv1.실패율-배열

jihyun-jeon committed:
feat: lv2.방문길이-구현

jihyun-jeon committed:
feat: lv2.올바른 괄호-스택

f-lab-jake commented on 프로그래머스/배열/lv1.모의고사-완전탐색.ts:
```diff
@@ -0,0 +1,108 @@
+    [3, 3, 1, 1, 2, 2, 4, 4, 5, 5],
+  ];
+
+  const scores = [0, 0, 0];
+
+  for (let i = 0; i < answers.length; i++) {
+    for (let j = 0; j < persons.length; j++) {
+      if (answers[i] === persons[j][i % persons[j].length]) {
+        scores[j]++;
+      }
+    }
+  }
```
`scores`를 수정하지마세요. hint: `map`을 쓰세요

f-lab-jake commented on 프로그래머스/스택/lv2.올바른괄호.ts:
```diff
@@ -0,0 +1,28 @@
+  const signs: Record<string, string> = { ")": "(" };
+
+  for (const str of s) {
+    const top = stack[stack.length - 1];
```
`at`쓰세요

f-lab-jake commented on 프로그래머스/배열/lv2.방문길이-구현.ts:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    const path1 = `${x}${y}${x2}${y2}`;
+    const path2 = `${x2}${y2}${x}${y}`;
```
`path`에 필요 이상의 자유도가 있습니다. 가령 대각선으로 이동하는 틀린 `path`도 포함해버릴수 있씁니다.
`d`를 포함시키세요.

jihyun-jeon committed:
refac: 5주차 코드 개선
- 모의고사 : 원본 변수값을 수정하지 않도록 개선
- 방문길이 : path에 방향 정보 포함시킴, 타입 개선
- 올바른괄호 : 배열의 마지막 요소 찾는 로직 at메서드 활용

jihyun-jeon commented:
피드백 반영하여 코드 수정했습니다!
[d4c2202](https://github.com/jihyun-jeon/ts-algorithm/pull/5/commits/d4c2202e6620682d8e182b27201f9626dba79d9c)

f-lab-jake commented on 프로그래머스/배열/lv2.방문길이-구현.ts:
```diff
@@ -33,14 +35,14 @@ function solution(dirs: string) {
 
-    const path1 = `${x}${y}${x2}${y2}`;
-    const path2 = `${x2}${y2}${x}${y}`;
+    const path1 = `${x}${y}${x2}${y2}${d}`;
+    const path2 = `${x2}${y2}${x}${y}${reverseD}`;
```
더 줄일수있습니다

f-lab-jake replied:
문자열안의 모든 변수들이 독립이 아니에요

jihyun-jeon replied:
path1,2의 중복을 제거하여
하나의 path 로 개선했습니다!
[2f7c4c6](https://github.com/jihyun-jeon/ts-algorithm/pull/5/commits/2f7c4c6d50bd6ae627f4099f39059ac1c36fb530)

jihyun-jeon committed:
refac: lv2.방문길이 코드 개선
: 방향성 판별위한 path1,path2 중복값 제거, 하나의 path로 판단하도록
: 방향 판별하는 함수 생성하여, 대각선인 경우 무시하도록 처리

jihyun-jeon committed:
refac: lv2.방문길이 코드 재개선
: path에 담길 값 최소화
: visit 자료구조 Set으로 수정

jihyun-jeon commented:
[67b8292](https://github.com/jihyun-jeon/ts-algorithm/pull/5/commits/67b82924914f6f597d6bf271fe0b6d4360536861)
코드 재개선 했습니다!

- path값을 `x,y,d` 구조로 단순화

- 역방향을 정방향과 동일한 경로로 처리
ex) 00 → 01 `U` , 01 → 00 `D`
      둘다 `00U` 라는 경로로 처리
